lib/internal/worker.js

```diff
@@ -0,0 +1,475 @@
+  }
+
+  [kOnErrorMessage](serialized) {
+    // This is what is called for uncaught exceptions.
+    const error = deserializeError(serialized);
+    this.emit('error', error);
+  }
+
+  [kOnMessage](message) {
+    switch (message.type) {
+      case 'upAndRunning':
+        return this.emit('online');
+      case 'couldNotSerializeError':
+        return this[kOnCouldNotSerializeErr]();
+      case 'errorMessage':
+        return this[kOnErrorMessage](message.error);
+      case 'stdioPayload':
+      {
+        const { stream, chunk, encoding } = message;
+        return this[kParentSideStdio][stream].push(chunk, encoding);
+      }
+      case 'stdioWantsMoreData':
+      {
+        const { stream } = message;
+        return this[kParentSideStdio][stream][kStdioWantsMoreDataCallback]();
+      }
+    }
+
+    assert.fail(`Unknown worker message type ${message.type}`);
+  }
+
+  [kDispose]() {
+    this[kHandle].onexit = null;
+    this[kHandle] = null;
+    this[kPort] = null;
+    this[kPublicPort] = null;
+
+    const { stdout, stderr } = this[kParentSideStdio];
+    this[kParentSideStdio] = null;
+
+    if (!stdout._readableState.ended) {
+      debug(`[${threadId}] explicitly closes stdout for ${this.threadId}`);
+      stdout.push(null);
+    }
+    if (!stderr._readableState.ended) {
+      debug(`[${threadId}] explicitly closes stderr for ${this.threadId}`);
+      stderr.push(null);
+    }
+  }
+
+  postMessage(...args) {
+    this[kPublicPort].postMessage(...args);
+  }
+
+  terminate(callback) {
+    if (this[kHandle] === null) return;
+
+    debug(`[${threadId}] terminates Worker with ID ${this.threadId}`);
+
+    if (typeof callback !== 'undefined')
+      this.once('exit', (exitCode) => callback(null, exitCode));
+
+    this[kHandle].stopThread();
+  }
+
+  ref() {
+    if (this[kHandle] === null) return;
+
+    this[kHandle].ref();
+    this[kPublicPort].ref();
+  }
+
+  unref() {
+    if (this[kHandle] === null) return;
+
+    this[kHandle].unref();
+    this[kPublicPort].unref();
+  }
+
+  get threadId() {
+    if (this[kHandle] === null) return -1;
+
+    return this[kHandle].threadId;
+  }
+
+  get stdin() {
+    return this[kParentSideStdio].stdin;
+  }
+
+  get stdout() {
+    return this[kParentSideStdio].stdout;
+  }
+
+  get stderr() {
+    return this[kParentSideStdio].stderr;
+  }
+}
+
+const workerStdio = {};
+if (!isMainThread) {
+  const port = getEnvMessagePort();
+  port[kWaitingStreams] = 0;
+  workerStdio.stdin = new ReadableWorkerStdio(port, 'stdin');
+  workerStdio.stdout = new WritableWorkerStdio(port, 'stdout');
+  workerStdio.stderr = new WritableWorkerStdio(port, 'stderr');
+}
+
+let originalFatalException;
+
+function setupChild(evalScript) {
+  // Called during bootstrap to set up worker script execution.
+  debug(`[${threadId}] is setting up worker child environment`);
+  const port = getEnvMessagePort();
+
+  const publicWorker = require('worker_threads');
+
+  port.on('message', (message) => {
```

Comment:
nit: on this initial iteration no need to do this now, but, as this progresses it would be good to eliminate the closure and make these normal top level functions.